src/main/java/com/loofah/graph/api/services/SkillService.java

```diff
@@ -0,0 +1,56 @@
+package com.loofah.graph.api.services;
+
+import com.loofah.graph.api.models.database.Skill;
+import com.loofah.graph.api.models.filters.SkillFilter;
+import com.loofah.graph.api.repositories.SkillRepository;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class SkillService {
+
+    private final SkillRepository skillRepository;
+
+    @Autowired
+    public SkillService(SkillRepository skillRepository) {
+        this.skillRepository = skillRepository;
+    }
+
+    public Skill getById(String id) {
+        return skillRepository.findById(id).get();
+    }
+
+    public List<Skill> getWithFilter(SkillFilter skillFilter) {
+        Query query = buildQuery(skillFilter);
+        return skillRepository.findWithQuery(query, Skill.class);
+    }
+
+    private Query buildQuery(SkillFilter skillFilter) {
+
+        Optional<SkillFilter> optionalFilter = Optional.ofNullable(skillFilter);
+
+        Query query = new Query();
+
+        optionalFilter.map(SkillFilter::getCategoryId).ifPresent(
+                categoryId -> query.addCriteria(Criteria.where("categoryId").is(categoryId))
+        );
+
+        optionalFilter.map(SkillFilter::getGradeId).ifPresent(
+                gradeId -> query.addCriteria(Criteria.where("gradeId").is(gradeId))
+        );
+
+        optionalFilter.map(SkillFilter::getCraftIds).ifPresent(
+                craftIds -> query.addCriteria(Criteria.where("craftIds").in(craftIds))
```

Comment:
If I’m following this logic correctly, it means we’ll match only skills for which their full set of associated  crafts are included in our search filter. So if we searched for backend skills, we would get all the backend only skills, but we wouldn’t get skills that are associated with both backend and devops. Is this the desired behaviour?